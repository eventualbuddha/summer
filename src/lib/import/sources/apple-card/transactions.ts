import { ImportedTransaction, ImportedTransactionKind } from '$lib/import/ImportedTransaction';
import { ParseStatementError } from '$lib/import/parse/errors';
import { MONEY_PATTERN, parseAmount } from '$lib/import/parse/money';
import type { PageTextLocation } from '$lib/import/statement/navigation';
import type { Statement } from '$lib/import/statement/Statement';
import { Result } from '@badrap/result';
import { DateTime } from 'luxon';

const MM_DD_YYYY_PATTERN = /^\s*\d{1,2}\/\d{1,2}\/\d{4}\s*$/;

export function* parseTransactions(
	statement: Statement
): Generator<Result<Transaction, ParseStatementError>> {
	const transactionSectionHeadings = statement.navigator.findAll(/Transactions by \w+/);

	for (const transactionsHeading of transactionSectionHeadings) {
		const dateTableHeaderLabel = transactionsHeading.findDown('Date', {
			alignment: 'left',
			maxAlignmentError: 2
		});

		if (!dateTableHeaderLabel) {
			yield Result.err(
				ParseStatementError.MissingLabel('Missing "Date" table header label', {
					searchFromText: transactionsHeading.text,
					searchDirection: 'down'
				})
			);
			return;
		}

		const firstDateValue = dateTableHeaderLabel.findDown(MM_DD_YYYY_PATTERN, {
			maxGap: 30,
			alignment: 'left',
			maxAlignmentError: 2
		});

		if (!firstDateValue) {
			yield Result.err(
				ParseStatementError.MissingValue('Missing first date value', {
					searchFromText: dateTableHeaderLabel.text,
					searchDirection: 'down'
				})
			);
			return;
		}

		let currentDateValue: PageTextLocation | undefined = firstDateValue.pageLocation;
		while (currentDateValue) {
			const transactionDate = DateTime.fromFormat(currentDateValue.text.str, 'MM/dd/yyyy');

			if (!transactionDate.isValid) {
				yield Result.err(
					ParseStatementError.InvalidValue('Invalid transaction date', {
						contentText: currentDateValue.text
					})
				);
				return;
			}

			const statementDescriptionValue = currentDateValue.findRight(/\w+/);

			if (!statementDescriptionValue) {
				yield Result.err(
					ParseStatementError.MissingValue('Missing transaction description', {
						searchFromText: currentDateValue.text,
						searchDirection: 'right'
					})
				);
				return;
			}

			const dailyCashAdjustmentLabel = statementDescriptionValue.findDown('Daily Cash Adjustment', {
				maxGap: 10,
				alignment: 'left'
			});

			if (dailyCashAdjustmentLabel) {
				const dailyCashPercentValue = dailyCashAdjustmentLabel.findRight(/\d+/);

				if (!dailyCashPercentValue) {
					yield Result.err(
						ParseStatementError.MissingValue('Missing daily cash percent value', {
							searchFromText: dailyCashAdjustmentLabel.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const dailyCashAmountValue = dailyCashPercentValue.findRight(MONEY_PATTERN);

				if (!dailyCashAmountValue) {
					yield Result.err(
						ParseStatementError.MissingValue('Missing daily cash amount value', {
							searchFromText: dailyCashPercentValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const parseDailyCashAmountResult = parseAmount(dailyCashAmountValue.text.str.trim());

				if (parseDailyCashAmountResult.isErr) {
					yield Result.err(
						ParseStatementError.InvalidValue('Invalid daily cash amount', {
							searchFromText: dailyCashAmountValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const dailyCashAmount = -parseDailyCashAmountResult.value;
				const amountValue = statementDescriptionValue.findRight(MONEY_PATTERN);

				if (!amountValue) {
					yield Result.err(
						ParseStatementError.MissingValue('Missing transaction amount value', {
							searchFromText: statementDescriptionValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const parseAmountResult = parseAmount(amountValue.text.str.trim());

				if (parseAmountResult.isErr) {
					yield Result.err(
						ParseStatementError.InvalidValue('Invalid transaction amount', {
							searchFromText: amountValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const amount = -parseAmountResult.value;

				const dailyCashPercent = parseFloat(dailyCashPercentValue.text.str.trim().replace('%', ''));
				const description = statementDescriptionValue.text.str.trim();
				const transaction = new Transaction(
					ImportedTransactionKind.Credit,
					transactionDate,
					amount,
					description,
					transactionsHeading.pageNumber,
					dailyCashPercent,
					dailyCashAmount
				);
				yield Result.ok(transaction);
			} else {
				const transactionTexts = currentDateValue.collectRight(/\w+/);

				if (transactionTexts.length !== 4) {
					yield Result.err(
						ParseStatementError.MissingValue('Unexpected transaction data', {
							searchFromText: currentDateValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const [descriptionValue, dailyCashPercentValue, dailyCashAmountValue, amountValue] =
					transactionTexts as [
						PageTextLocation,
						PageTextLocation,
						PageTextLocation,
						PageTextLocation
					];

				const parseDailyCashAmountResult = parseAmount(dailyCashAmountValue.text.str.trim());

				if (parseDailyCashAmountResult.isErr) {
					yield Result.err(
						ParseStatementError.InvalidValue('Invalid daily cash amount', {
							searchFromText: dailyCashAmountValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const dailyCashAmount = parseDailyCashAmountResult.value;
				const dailyCashPercent = parseFloat(dailyCashPercentValue.text.str.trim().replace('%', ''));
				const parseAmountResult = parseAmount(amountValue.text.str.trim());

				if (parseAmountResult.isErr) {
					yield Result.err(
						ParseStatementError.InvalidValue('Invalid transaction amount', {
							searchFromText: amountValue.text,
							searchDirection: 'right'
						})
					);
					return;
				}

				const amount = -parseAmountResult.value;
				const description = descriptionValue.text.str.trim();

				const transaction = new Transaction(
					ImportedTransactionKind.Charge,
					transactionDate,
					amount,
					description,
					transactionsHeading.pageNumber,
					dailyCashPercent,
					dailyCashAmount
				);

				yield Result.ok(transaction);
			}

			currentDateValue = currentDateValue.findDown(MM_DD_YYYY_PATTERN, {
				maxGap: 30,
				alignment: 'left'
			});
		}
	}
}

export class Transaction extends ImportedTransaction {
	readonly dailyCashPercent: number;
	readonly dailyCashAmount: number;

	constructor(
		kind: ImportedTransactionKind,
		date: DateTime,
		amount: number,
		description: string,
		pageNumber: number,
		dailyCashPercent: number,
		dailyCashAmount: number
	) {
		super(kind, date, amount, description, pageNumber);
		this.dailyCashPercent = dailyCashPercent;
		this.dailyCashAmount = dailyCashAmount;
	}
}
