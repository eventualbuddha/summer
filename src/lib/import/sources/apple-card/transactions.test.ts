import { expect, test } from 'bun:test';
import { parseTransactions } from './transactions';
import { apple } from '../../../../../tests/fixtures';

test('parse transactions', () => {
	expect(
		parseTransactions(apple.statement())
			.toArray()
			.map((r) => r.unwrap())
	).toMatchSnapshot();
});
