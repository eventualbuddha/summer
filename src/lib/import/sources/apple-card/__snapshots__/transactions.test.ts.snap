// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse transactions 1`] = `
[
  Transaction {
    "amount": -150,
    "dailyCashAmount": 2,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1735776000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "CITY OF ALAMEDA IPS 2263 SANTA CLARA AVE 202 ALAMEDA 94501 CA USA",
  },
  Transaction {
    "amount": -300,
    "dailyCashAmount": 3,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 4,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1735948800000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "OAKLAND PARK METER IPS250 FRANK OGAWA PLZ 6300 OAKLAND 94612 CA USA",
  },
  Transaction {
    "amount": -1299,
    "dailyCashAmount": 39,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 9,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736380800000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -999,
    "dailyCashAmount": 30,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 12,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736640000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -1499,
    "dailyCashAmount": 45,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 12,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736640000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -1599,
    "dailyCashAmount": 48,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 13,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736726400000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -1499,
    "dailyCashAmount": 45,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 13,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736726400000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -20600,
    "dailyCashAmount": 206,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 14,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736812800000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "LONGBOTTOM & CO. DDS 291 GROVE ST OXFORDSHIRE 29384 UK",
  },
  Transaction {
    "amount": -1017,
    "dailyCashAmount": 10,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 16,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736985600000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "YOGOFINA OXFORD 3412 PARK AVE OXFORD 84710 UK",
  },
  Transaction {
    "amount": -999,
    "dailyCashAmount": 30,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 21,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737417600000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -799,
    "dailyCashAmount": 24,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 21,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737417600000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 3,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -1898,
    "dailyCashAmount": 57,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 22,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737504000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY CUPERTINO 95014 CA USA",
  },
  Transaction {
    "amount": -25730,
    "dailyCashAmount": 257,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 22,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737504000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "FLOURISH AND BLOTTSCO2000 GRAND STREET 0571839400 89205 UK",
  },
  Transaction {
    "amount": -7240,
    "dailyCashAmount": 145,
    "dailyCashPercent": 2,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 25,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737763200000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "THE THREE BROOMSTICKS KITCHEN AND BAR HOGSMEADE UK",
  },
  Transaction {
    "amount": -349,
    "dailyCashAmount": 10,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 28,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1738022400000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -1499,
    "dailyCashAmount": 45,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 28,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1738022400000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK CUPERTINO 95014 CA USA",
  },
  Transaction {
    "amount": -4999,
    "dailyCashAmount": 50,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 29,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1738108800000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "MUGGLE MANIA MADNESS INC. CITY OF LONDON UK",
  },
  Transaction {
    "amount": -3795,
    "dailyCashAmount": 114,
    "dailyCashPercent": 3,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 30,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1738195200000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "APPLE.COM/BILL ONE APPLE PARK WAY 866-712-7753 95014 CA USA",
  },
  Transaction {
    "amount": -100,
    "dailyCashAmount": 1,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 4,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1735948800000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "OAKLAND PARK METER IPS250 FRANK OGAWA PLZ 6300 OAKLAND 94612 CA USA",
  },
  Transaction {
    "amount": -10700,
    "dailyCashAmount": 107,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 6,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736121600000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "COUNTY SPECIALTY POTIONS CRESTMORE AVE GODRICS HOLLOW UK",
  },
  Transaction {
    "amount": -150,
    "dailyCashAmount": 2,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 16,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1736985600000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "CITY OF ALAMEDA IPS 2263 SANTA CLARA AVE 202 ALAMEDA 94501 CA USA",
  },
  Transaction {
    "amount": -10700,
    "dailyCashAmount": 107,
    "dailyCashPercent": 1,
    "date": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 27,
        "hour": 0,
        "millisecond": 0,
        "minute": 0,
        "month": 1,
        "second": 0,
        "year": 2025,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1737936000000,
      "weekData": null,
    },
    "kind": "Charge",
    "pageNumber": 4,
    "statementDescription": "COUNTY SPECIALTY POTIONS CRESTMORE AVE GODRICS HOLLOW UK",
  },
]
`;
